{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b3a7c761-1f75-49fb-aa15-f307e93e4211",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from Bio import motifs\n",
    "from Bio import SeqIO\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ed607c7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "# Define the PPM function\n",
    "def ppm(C):\n",
    "    return C\n",
    "\n",
    "# Define the S function, B is a background\n",
    "def s(C, B):\n",
    "    return np.log2(ppm(C) / B)\n",
    "\n",
    "# Define the ICtotal function\n",
    "def ic_total(C):\n",
    "    return np.log2(len(C))\n",
    "\n",
    "# Define the U function\n",
    "def u(C):\n",
    "    return -np.sum(ppm(C) * np.log2(ppm(C)), axis=0)\n",
    "\n",
    "# Define the ICfinal function\n",
    "def ic_final(C):\n",
    "    return ic_total(C) - u(C)\n",
    "\n",
    "# Define the IC function\n",
    "def ic(C):\n",
    "    return ppm(C) * ic_final(C)\n",
    "\n",
    "# Calculate IC values\n",
    "def calc_ic_df(pfm_df):\n",
    "    B = 0.25\n",
    "    ppm_df = pfm_df.apply(ppm, axis=1)\n",
    "    s_df = ppm_df.apply(lambda x: s(x, B))\n",
    "    pfm_df\n",
    "    ic_df = pfm_df.apply(ic, axis=1)\n",
    "    ic_df = ic_df.transpose()\n",
    "    ic_t = ic_df.sum(axis=0)\n",
    "    ic_t = ic_t.round(2)\n",
    "    return ic_t.to_list()\n",
    "\n",
    "def trim_motif (motif, ic_t, reverse, threshold = 0.5):\n",
    "\n",
    "    began = False\n",
    "    new_motif = \"\"\n",
    "    ran = range(0)\n",
    "    if reverse:\n",
    "        ran = reversed(range(len(motif)))\n",
    "    else:\n",
    "        ran = range(0, len(motif))\n",
    "    new_ic = []\n",
    "    for i in ran:\n",
    "        if ic_t[i] > threshold:\n",
    "            began = True\n",
    "            new_motif += motif[i]\n",
    "            new_ic.append(ic_t[i])\n",
    "            began = True\n",
    "        elif ic_t[i] <= threshold and began:\n",
    "            new_motif += motif[i]\n",
    "            new_ic.append(ic_t[i])\n",
    "    \n",
    "    # trim from back as well\n",
    "    if reverse:\n",
    "        return new_motif[::-1], list(reversed(new_ic))\n",
    "\n",
    "    return trim_motif(new_motif, new_ic, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "41c6e90e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# dict storing rules for combining two IUPAC letters\n",
    "IUPAC_TO_IUPAC = {\n",
    "        \"A\" : {\"A\":\"A\", \"C\":\"M\", \"G\":\"R\", \"T\":\"W\", \"M\":\"M\", \"R\":\"R\", \"W\":\"W\", \"S\":\"V\", \"Y\":\"H\", \"K\":\"D\", \"V\":\"V\", \"H\":\"H\", \"D\":\"D\", \"B\":\"N\", \"N\":\"N\",},\n",
    "        \"C\" : {\"A\":\"M\", \"C\":\"C\", \"G\":\"S\", \"T\":\"Y\", \"M\":\"M\", \"R\":\"V\", \"W\":\"H\", \"S\":\"S\", \"Y\":\"Y\", \"K\":\"B\", \"V\":\"V\", \"H\":\"H\", \"D\":\"N\", \"B\":\"B\", \"N\":\"N\",},\n",
    "        \"G\" : {\"A\":\"R\", \"C\":\"S\", \"G\":\"G\", \"T\":\"K\", \"M\":\"V\", \"R\":\"R\", \"W\":\"D\", \"S\":\"S\", \"Y\":\"B\", \"K\":\"K\", \"V\":\"V\", \"H\":\"N\", \"D\":\"D\", \"B\":\"B\", \"N\":\"N\",},\n",
    "        \"T\" : {\"A\":\"W\", \"C\":\"Y\", \"G\":\"K\", \"T\":\"T\", \"M\":\"H\", \"R\":\"D\", \"W\":\"W\", \"S\":\"B\", \"Y\":\"Y\", \"K\":\"K\", \"V\":\"N\", \"H\":\"H\", \"D\":\"D\", \"B\":\"B\", \"N\":\"N\",},\n",
    "        \"M\" : {\"A\":\"M\", \"C\":\"M\", \"G\":\"V\", \"T\":\"H\", \"M\":\"M\", \"R\":\"V\", \"W\":\"H\", \"S\":\"V\", \"Y\":\"H\", \"K\":\"N\", \"V\":\"V\", \"H\":\"H\", \"D\":\"N\", \"B\":\"N\", \"N\":\"N\",},\n",
    "        \"R\" : {\"A\":\"R\", \"C\":\"V\", \"G\":\"R\", \"T\":\"D\", \"M\":\"V\", \"R\":\"R\", \"W\":\"D\", \"S\":\"V\", \"Y\":\"N\", \"K\":\"D\", \"V\":\"V\", \"H\":\"N\", \"D\":\"D\", \"B\":\"N\", \"N\":\"N\",},\n",
    "        \"W\" : {\"A\":\"W\", \"C\":\"H\", \"G\":\"D\", \"T\":\"W\", \"M\":\"H\", \"R\":\"D\", \"W\":\"W\", \"S\":\"N\", \"Y\":\"H\", \"K\":\"D\", \"V\":\"N\", \"H\":\"H\", \"D\":\"D\", \"B\":\"N\", \"N\":\"N\",},\n",
    "        \"S\" : {\"A\":\"V\", \"C\":\"S\", \"G\":\"S\", \"T\":\"B\", \"M\":\"V\", \"R\":\"V\", \"W\":\"N\", \"S\":\"S\", \"Y\":\"B\", \"K\":\"B\", \"V\":\"V\", \"H\":\"N\", \"D\":\"N\", \"B\":\"B\", \"N\":\"N\",},\n",
    "        \"Y\" : {\"A\":\"H\", \"C\":\"Y\", \"G\":\"B\", \"T\":\"Y\", \"M\":\"H\", \"R\":\"N\", \"W\":\"H\", \"S\":\"B\", \"Y\":\"Y\", \"K\":\"B\", \"V\":\"N\", \"H\":\"H\", \"D\":\"N\", \"B\":\"B\", \"N\":\"N\",},\n",
    "        \"K\" : {\"A\":\"D\", \"C\":\"B\", \"G\":\"K\", \"T\":\"K\", \"M\":\"N\", \"R\":\"D\", \"W\":\"D\", \"S\":\"B\", \"Y\":\"B\", \"K\":\"K\", \"V\":\"N\", \"H\":\"N\", \"D\":\"D\", \"B\":\"B\", \"N\":\"N\",},\n",
    "        \"V\" : {\"A\":\"V\", \"C\":\"V\", \"G\":\"V\", \"T\":\"N\", \"M\":\"V\", \"R\":\"V\", \"W\":\"N\", \"S\":\"V\", \"Y\":\"N\", \"K\":\"N\", \"V\":\"V\", \"H\":\"N\", \"D\":\"N\", \"B\":\"N\", \"N\":\"N\",},\n",
    "        \"H\" : {\"A\":\"H\", \"C\":\"H\", \"G\":\"N\", \"T\":\"H\", \"M\":\"H\", \"R\":\"N\", \"W\":\"H\", \"S\":\"N\", \"Y\":\"H\", \"K\":\"N\", \"V\":\"N\", \"H\":\"H\", \"D\":\"N\", \"B\":\"N\", \"N\":\"N\",},\n",
    "        \"D\" : {\"A\":\"D\", \"C\":\"N\", \"G\":\"D\", \"T\":\"D\", \"M\":\"N\", \"R\":\"D\", \"W\":\"D\", \"S\":\"N\", \"Y\":\"N\", \"K\":\"D\", \"V\":\"N\", \"H\":\"N\", \"D\":\"D\", \"B\":\"N\", \"N\":\"N\",},\n",
    "        \"B\" : {\"A\":\"N\", \"C\":\"B\", \"G\":\"B\", \"T\":\"B\", \"M\":\"N\", \"R\":\"N\", \"W\":\"N\", \"S\":\"B\", \"Y\":\"B\", \"K\":\"B\", \"V\":\"N\", \"H\":\"N\", \"D\":\"N\", \"B\":\"B\", \"N\":\"N\",},\n",
    "        \"N\" : {\"A\":\"N\", \"C\":\"N\", \"G\":\"N\", \"T\":\"N\", \"M\":\"N\", \"R\":\"N\", \"W\":\"N\", \"S\":\"N\", \"Y\":\"N\", \"K\":\"N\", \"V\":\"N\", \"H\":\"N\", \"D\":\"N\", \"B\":\"N\", \"N\":\"N\",},\n",
    "    }\n",
    "def motif_contains(m1, m2):\n",
    "    \"\"\"\n",
    "    Checks if IUPAC motif m2 is identically contained in IUPAC motif m1.\n",
    "    Returns a tuple (idx, ident), where\n",
    "    idx is the first index of m2 in m1 if m1 is contained in m2 and None if it is not contained\n",
    "    ident is True if the substring of m1 at idx of length len(m2) is identical to m2\n",
    "    \n",
    "    Examples:\n",
    "    ATC is identically contained in CATC at index 1\n",
    "    ATC is non-identically contained in CATS at index 1\n",
    "    ATN is not contained in CATS\n",
    "    \"\"\"\n",
    "    lm1, lm2 = len(m1), len(m2)\n",
    "    if lm2 > len(m1):\n",
    "        return False\n",
    "    for i in range(lm1 - lm2 + 1):\n",
    "        identical = True\n",
    "        for j in range(lm2):\n",
    "            if m1[i+j] != m2[j]:\n",
    "                identical = False\n",
    "            if m2[j] != 'N':\n",
    "                if m1[i+j] == 'N':\n",
    "                    # do not match agaist TypeI motif gaps\n",
    "                    break\n",
    "                if m1[i+j] != IUPAC_TO_IUPAC[m1[i+j]][m2[j]]:\n",
    "                    break\n",
    "        else: # if not break\n",
    "            return i, identical\n",
    "    return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03ca5de2",
   "metadata": {},
   "outputs": [],
   "source": [
    "iupac = {'N': 'ATCG', \n",
    "         'V' : 'ACG', 'H' : 'ACT', 'D' : 'AGT', 'B' : 'CGT', \n",
    "         'M' : 'AC', 'K' : 'GT', 'W' : 'AT', 'S' : 'GC', 'Y' : 'CT', 'R' : 'AG',\n",
    "         'A' : 'A', 'T' : 'T', 'G' : 'G', 'C' : 'C'}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57c1b228",
   "metadata": {},
   "source": [
    "## postprocessing of PWMs of Xylella Nanopore sequencing data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5314ab30",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_files_dict = dict()\n",
    "for dataname in os.listdir('../../samples'):\n",
    "    for suffix in ['_peaks_uBH_0.001_peak_dist_2_min_cov_20_min_dist_20_k_3_kmer_quantile_0.25',\n",
    "                '_peaks_uBH_0.001_peak_dist_2_min_cov_20_min_dist_20',\n",
    "                '_peaks_uBH_0.001']:\n",
    "        fasta_files_dict[dataname + suffix] = '../../samples/' + dataname + '/peaks/' + dataname + suffix + '.fasta'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb948370",
   "metadata": {},
   "outputs": [],
   "source": [
    "for filename in os.listdir('training_xylella'):\n",
    "    filepath = 'training_xylella' + filename\n",
    "    fasta_file = fasta_files_dict[filename]\n",
    "    csv_file = 'motifs_xylella/' + filename + '.csv'\n",
    "    pwm_directory = filepath + '/recover_PWM_500'\n",
    "\n",
    "    # trim pwm based on IC and filter out motifs shorter than 3 nt\n",
    "    motif_list_trim = list()\n",
    "    for filename in os.listdir(pwm_directory):\n",
    "        f = os.path.join(pwm_directory, filename)\n",
    "        pwm = pd.read_csv(f, delimiter=\" \", names=[\"A\",\"C\",\"G\",\"T\"])\n",
    "        motif = motifs.Motif(alphabet=\"ACGT\", counts=pwm)\n",
    "        motif_trim = trim_motif(str(motif.degenerate_consensus), calc_ic_df(pwm), False, 1.5)[0]\n",
    "        if motif_trim and len(motif_trim) > 3:\n",
    "            motif_list_trim.append(motif_trim)\n",
    "    motif_list_trim = list(set(motif_list_trim))\n",
    "\n",
    "    # filter out motifs that are represented by other more unspecific motifs\n",
    "    motif_list_combined = list()\n",
    "    for motif1 in motif_list_trim:\n",
    "        flag = True\n",
    "        for motif2 in motif_list_trim:\n",
    "            if motif1 != motif2:\n",
    "                cont = motif_contains(motif1,motif2)\n",
    "                if cont and cont[1]:\n",
    "                    print(motif1,motif2)\n",
    "                    flag = False\n",
    "                    break\n",
    "        if flag:\n",
    "            motif_list_combined.append(motif1)\n",
    "    motif_list_combined = list(set(motif_list_combined))\n",
    "\n",
    "    # count occurrences of motifs\n",
    "    motif_counts = []\n",
    "    for motif in motif_list_combined:\n",
    "        motif_count = 0\n",
    "        for rec in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "            regex = r\"\"\n",
    "            for cc in motif:\n",
    "                regex += f\"[{iupac[cc]}]\"\n",
    "            motif_count += len(re.findall(re.compile(regex, re.IGNORECASE), str(rec.seq)))\n",
    "        motif_counts.append(motif_count)\n",
    "\n",
    "    df = pd.DataFrame({'motif' : motif_list_combined, 'frequency' : motif_counts})\n",
    "    df.to_csv(f\"{csv_file}.csv\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad8f4452",
   "metadata": {},
   "source": [
    "## postprocessing of PWMs of artificial datasets with controlled TPR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "616719ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_files_dict = dict()\n",
    "for dataname in os.listdir('../../../artificial_Riv19/'):\n",
    "    fasta_files_dict[dataname] = '../../../artificial_Riv19/' + dataname + '.fasta'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1af3b881",
   "metadata": {},
   "outputs": [],
   "source": [
    "for filename in os.listdir('training_artificial'):\n",
    "    filepath = 'training_artificial' + filename\n",
    "    fasta_file = fasta_files_dict[filename]\n",
    "    csv_file = 'motifs_artificial/' + filename + '.csv'\n",
    "    pwm_directory = filepath + '/recover_PWM_500'\n",
    "\n",
    "    # trim pwm based on IC and filter out motifs shorter than 3 nt\n",
    "    motif_list_trim = list()\n",
    "    for filename in os.listdir(pwm_directory):\n",
    "        f = os.path.join(pwm_directory, filename)\n",
    "        pwm = pd.read_csv(f, delimiter=\" \", names=[\"A\",\"C\",\"G\",\"T\"])\n",
    "        motif = motifs.Motif(alphabet=\"ACGT\", counts=pwm)\n",
    "        motif_trim = trim_motif(str(motif.degenerate_consensus), calc_ic_df(pwm), False, 1.5)[0]\n",
    "        if motif_trim and len(motif_trim) > 3:\n",
    "            motif_list_trim.append(motif_trim)\n",
    "    motif_list_trim = list(set(motif_list_trim))\n",
    "\n",
    "    # filter out motifs that are represented by other more unspecific motifs\n",
    "    motif_list_combined = list()\n",
    "    for motif1 in motif_list_trim:\n",
    "        flag = True\n",
    "        for motif2 in motif_list_trim:\n",
    "            if motif1 != motif2:\n",
    "                cont = motif_contains(motif1,motif2)\n",
    "                if cont and cont[1]:\n",
    "                    print(motif1,motif2)\n",
    "                    flag = False\n",
    "                    break\n",
    "        if flag:\n",
    "            motif_list_combined.append(motif1)\n",
    "    motif_list_combined = list(set(motif_list_combined))\n",
    "\n",
    "    # count occurrences of motifs\n",
    "    motif_counts = []\n",
    "    for motif in motif_list_combined:\n",
    "        motif_count = 0\n",
    "        for rec in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "            regex = r\"\"\n",
    "            for cc in motif:\n",
    "                regex += f\"[{iupac[cc]}]\"\n",
    "            motif_count += len(re.findall(re.compile(regex, re.IGNORECASE), str(rec.seq)))\n",
    "        motif_counts.append(motif_count)\n",
    "\n",
    "    df = pd.DataFrame({'motif' : motif_list_combined, 'frequency' : motif_counts})\n",
    "    df.to_csv(f\"{csv_file}.csv\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "comp-methods",
   "language": "python",
   "name": "comp-methods-env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
